elixir/apps/web/lib/web/live/gateway_groups/new.ex

```diff
@@ -128,4 +124,35 @@ defmodule Web.GatewayGroups.New do
 
     {:noreply, socket}
   end
+
+  defp docker_command(secret) do
+    """
+    docker run -d \\
```

Comment:
@AndrewDryga minor correction -- We don't want to override `RUST_LOG` by default -- #2477 will set a sensible default for production there.

For relay it will be:

```
docker run -d \
  --restart=always \
  --health-cmd="lsof -i UDP | grep firezone-relay" \
  --name=firezone-relay \
  --network host \
  --env TOKEN="..." \
  ghcr.io/firezone/relay:${FIREZONE_VERSION:-1}
```

On Linux I think we *could* map the port range for the relay explicitly instead of using `--network host`, but if I'm wrong then someone will run that command and it'll crash their machine 🤠

Reply:
Actually, on second thought, the user experience here is broken due to #2482 

Perhaps it makes sense to add a `--env ID=$(uuidgen) \` line to these?

Reply:
That will still require connlib changes to pick that up, just FYI.

Reply:
Yeah, the fact that #2482 was still open is why I didn't really bother with the contents of the command and just focused on making sure it formatted properly and didn't have any HTML tags in the copied code.

Reply:
@bmanifold sgtm we can revisit the contents then later

Reply:
@jamilbk we need UUID to be generated by the `rust` and saved to a well-known location, `ID=$(uuidgen)` will result in a new ID every time the docker runtime restarts.

Reply:
> saved to a well-known location

The issue is that we've been operating under the assumption Gateways don't require persistence storage, so I've been communicating that architecture to early access users for a couple months now. I'm not opposed to requiring volume mounts for Gateways, we just need to write documentation to that effect -- upgrades, backups, restores need to make sure that file is saved.

> ID=$(uuidgen) will result in a new ID every time the docker runtime restarts.

`\`uuidgen\`` seems to do the trick! And saves the user from having to mount a volume.

```
jamil@Jamils-MacBook-Air:~ % docker run --env ID=`uuidgen` --restart=always alpine /bin/sh -c 'echo $ID'
A2BB2590-27AB-421A-B98C-1D925792E1F0

jamil@Jamils-MacBook-Air:~ % docker logs d2b54faad844
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
A2BB2590-27AB-421A-B98C-1D925792E1F0
```

Reply:
The other thing I worry about making this a file is that when the VM is cloned to replicate the Gateway, we'll have Gateways upserting over each other.

Reply:
> The issue is that we've been operating under the assumption Gateways don't require persistence storage, so I've been communicating that architecture to early access users for a couple months now. I'm not opposed to requiring volume mounts for Gateways, we just need to write documentation to that effect -- upgrades, backups, restores need to make sure that file is saved.

We don't *require* persistent storage, we don't need backups or anything like that. We just have to write it somewhere so that it survives reboots of the docker runtime. Your example is not correct in this regard:
```
❯ docker run --env ID=`uuidgen` alpine /bin/sh -c 'echo $ID'
B16DF55B-A56C-4B0D-8F54-BE1230AA88A9

❯ docker run --env ID=`uuidgen` alpine /bin/sh -c 'echo $ID'
07CC17DB-4A77-42F7-9998-E4B673B0EBA0

❯ docker run --env ID=`uuidgen` alpine /bin/sh -c 'echo $ID'
DC6C703A-3F56-4935-9F49-50170868CDE9
```

`uuidgen ` will give new results every time the host reboots, the container runtime reboots (so the script is re-evaluated), etc.

The solution is to cache the result in a file and accept the environment variable at the same time. So that `connlib`:
1. reads `/etc/firezone/.instance_id` file;
2. reads `FIREZONE_INSTANCE_ID` environment variable (notice `FIREZONE_` prefix because we will provide binaries to run outside of Docker so system env can be polluted);
3. reads SMBios;
and whatever the result is - write it to `/etc/firezone/.instance_id` file.

We also want to attach a docker volume to our docker script:
```
--volume /etc/firezone
```

This will provide the maximum level of persistence: we will reuse IDs if they are set explicitly and do our best to save generated values, but still work as before if everything fails.

Reply:
> The other thing I worry about making this a file is that when the VM is cloned to replicate the Gateway, we'll have Gateways upserting over each other (though I suppose that would happen too for the Docker case)

That is exactly why we should NOT ask users to have persistence, backups of anything else. Losing ID is fine, if not too often.

Reply:
@AndrewDryga Losing ID is a usability problem because new Gateways then show up in the admin portal UI with different authorizations and logs, which is confusing to admins and security teams who are treating Gateways as infrastructure elements.

It's also a problem if we plan to charge per Gateway, which we are thinking of doing (cc @jefferenced). Most organizations will have only a handful of Gateway Groups with 2 or 3 Gateways in them each.

Based on the above, asking users to persist an ID file in addition to managing a token seems suboptimal.

Why allow upserting Gateways at all? If we can generate new tokens for an existing gateway group, can't we just create a new Gateway when a user requests it, embed its UUID in the new token, and then identify and authenticate with one token?

No one has asked for autoscaling Gateways, nor is it how other products in the space work, so I'm perplexed why we are building it.

Reply:
Just to get this working now, what if we generate a UUID in the portal when we show the command? The ID would be persisted just like the TOKEN within the docker command itself:

```
ID=<random_uuid_generated_by_portal>
TOKEN=<gateway_group_token>
```

We can also show the ID in the portal for that particular Gateway, allowing the user to reuse that instance to resume in case an `.instance_id` file is lost.

But now we're back to just encoding the ID in token itself and just sticking to one thing for the users to maintain.

Reply:
> Losing ID is a usability problem because new Gateways then show up in the admin portal UI with different authorizations and logs, which is confusing to admins and security teams who are treating Gateways as infrastructure elements.

Exactly. That's why I think we need to additionally persist them to a file. With the scheme provided above they will be as persistent as possible, but gateways are ephemeral and it's an important tradeoff we took months ago.

> It's also a problem if we plan to charge per Gateway, which we are thinking of doing (cc @jefferenced). Most organizations will have only a handful of Gateway Groups with 2 or 3 Gateways in them each.

Why do we want to do it per gateway, not per gateway group?

> Why allow upserting Gateways at all? If we can generate new tokens for an existing gateway group, can't we just create a new Gateway when a user requests it, embed its UUID in the new token, and then identify and authenticate with one token?

This will make the experience of scaling the gateways so much more complicated (deploying 3 - copy-paste 3 times) and will complicate DevOps stuff for our clients a lot. Instead of one Google Instance Group/AWS Auto-Scaling Group/K8s Replica Set they will need to create 3: each with its own environment variable.

> No one has asked for autoscaling Gateways, nor is it how other products in the space work, so I'm perplexed why we are building it.

We were designing a solution where you can start "the old way" and then go as much zero-trust as possible. 

Current design decisions support both of them. 

If you need a "bastion gateway" you can deploy one or more replicas of a gateway to your environment and allow them to access your resources, fine-grain controlling which user can access certain resources in the Firezone UI.

If you want a gateway to be deployed as close to the resource as possible (zero-trust) you most likely will end up with a sidecar container, which means that if your application needs autoscaling, or at least some form of manually-initiated scaling, so Firezone needs to support that too. Especially because for us this costs nothing: allowing to reuse of a single token between multiple gateways.

Reply:
> Just to get this working now, what if we generate a UUID in the portal when we show the command? The ID would be persisted just like the TOKEN within the docker command itself:

This means we will have a token per gateway, which I feel would be a really bad design and usability decision for the reasons described above.